### PR TITLE
tests: net: stop whitelisting

### PR DESCRIPTION
**Pull request by tyrantlizard77**
This is bad practice and may conceal issues on platforms
not in the whitelist.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

File: tests/net/lib/mqtt_publisher/testcase.yaml [hunk 1]
```diff
@@ -2,9 +2,8 @@ common:
   tags: net mqtt
   harness: net
   depends_on: netif
-  platform_whitelist: native_posix qemu_x86 qemu_cortex_m3
+  min_ram: 16
 tests:
   net.mqtt:
```
Comment (nashif) [inline tests/net/lib/mqtt_publisher/testcase.yaml]: you should at least have 1 key in this section, thats why it fails.

**Review by nashif — CHANGES_REQUESTED** [review]

**Comment by jukkar** [comment]
I do not mind if we run the tests with more platforms. The original reason the whitelist looks like this, was that the sanity took very long time to finish and as the tests are basically hw agnostic, we limited the number of platforms that run the tests.

**Comment by andrewboie** [comment]
> The original reason the whitelist looks like this, was that the sanity took very long time to finish and as the tests are basically hw agnostic, we limited the number of platforms that run the tests.

Going forward please do not do this. See #15883 #15878 #15881

**Comment by nashif** [comment]
@jukkar 

> The original reason the whitelist looks like this, was that the sanity took very long time to finish and as the tests are basically hw agnostic, we limited the number of platforms that run the tests.

If I remember correctly, the whitelisting happened when we started hitting all kind of issues when you run those tests on some boards (the atmel board). 

@andrewboie 

> Going forward please do not do this.

Whitelisting should be fine if you are writing a test for a certain configuration and nothing else, for example something that would use the network capabilities of native_posix to test a protocol. In that case we are not testing the platform, but some upper software layer or API. Trying to make those tests run on everything does not make sense. Think of it as a replacement to a unit test, something we do not have in zephyr or is not being used widely given the nature of the project and the need for OS primitives for almost everything we do.

**Comment by andrewboie** [comment]
> Whitelisting should be fine if you are writing a test for a certain configuration and nothing else, for example something that would use the network capabilities of native_posix to test a protocol. In that case we are not testing the platform, but some upper software layer or API. Trying to make those tests run on everything does not make sense. Think of it as a replacement to a unit test, something we do not have in zephyr or is not being used widely given the nature of the project and the need for OS primitives for almost everything we do.

Given that I've found bugs I can't agree with this.
We want a working network stack on x86_64, right?
We want to test the network stack on an ARM target that enables user mode, correct?
I'm about to add an qemu_x86_coverage target which enables CONFIG_COVERAGE=y to stop the neverending tide of bit-rot when it comes to code coverage, and net tests won't be built with that either!

**Comment by andrewboie** [comment]
Found another one: #15884

**Comment by andrewboie** [comment]
> If I remember correctly, the whitelisting happened when we started hitting all kind of issues when you run those tests on some boards (the atmel board).

So let's fix them?

**Comment by nashif** [comment]
> Given that I've found bugs I can't agree with this.

you do not agree with what? I actually agree with this change in principle.

I am saying that there are cases where tests are designed to be run with one and a special configuration, it should not be a catch-all rule that everything need to be built and run on every board. Some cases will NOT work on real hardware and the test will be isolated to limited functionality in unitest fashion and not the complete solution.

We will always find build issues or stack overflows because boards are different in many ways, there are different networking interfaces being used etc. etc. The combination should also be tested of course but you can't cover every level of software testing on all boards and all combination of configs of how you connect to the network in one test.

**Comment by andrewboie** [comment]
> I am saying that there are cases where tests are designed to be run with one and a special configuration, it should not be a catch-all rule that everything need to be built and run on every board. Some cases will NOT work on real hardware and the test will be isolated to limited functionality in unitest fashion and not the complete solution.

We're probably in violent agreement.
Yes, if a test is intended for one board, one config, sure whitelist it.
I don't think that applies to any of the tests here however.
I did intentionally leave the whitelist in for tests/net/all

**Comment by jukkar** [comment]
> With this patch, "sanitycheck --all -t net" is only 1400 tests so you're really not saving much and instead are just sweeping issues under the rug

With current master:
* `sanitycheck -t net` -> 253 tests
* `sanitycheck -t net --all`-> 746 tests

And with this PR:
* `sanitycheck -t net` -> 335 tests
* `sanitycheck -t net --all`-> 1424 tests

Without --all option the difference is not much but with --all it is quite a lot.

**Comment by andrewboie** [comment]
> Without --all option the difference is not much but with --all it is quite a lot.

~1400 is still not a lot.
And I've found many issues already.
This whitelisting is incredibly misguided.

**Comment by nashif** [comment]
our daily build with --all is not much faster and can deal with the additional builds, however, we really need to start classifying those tests, I am sure I will hit lots of issues if I run them all on real hardware now, especially the e70

**Review by nashif — APPROVED** [review]

**Comment by pfalcon** [comment]
> Xtensa? Nios2? ARC? RiscV?

Yes, those need to be blacklisted instead, in case of issues, which are many in some areas (e.g. https://github.com/zephyrproject-rtos/zephyr/pull/12984).

---

+1 from me, if we have CI capacity for it. And I'd be surprised to hear that we have, given the changes like 11aa39bc2591d62616e185a712a51bd2d1db6290 . There, for breadcrumb savings in CI runtime (say, run 1000 tests instead of 1005), a noticeable test organization/maintenance regression was made. Here, we say that running 1400 instead of 700 tests (100% more) is ok.

**Comment by andrewboie** [comment]
> > Xtensa? Nios2? ARC? RiscV?
> 
> Yes, those need to be blacklisted instead, in case of issues, which are many in some areas (e.g. #12984).

What does #12984 have to do with this? Did you mean to link something else? If the network stack has numerous issues on non-x86/ARM where are the open bugs?

We don't need to blacklist anything, I just ran this PR with "sanitycheck --all -t net" and everything passed except 3 tests on mimxrt1050_evk so ¯\_(ツ)_/¯
 
> +1 from me, if we have CI capacity for it. And I'd be surprised to hear that we have, given the changes like [11aa39b](https://github.com/zephyrproject-rtos/zephyr/commit/11aa39bc2591d62616e185a712a51bd2d1db6290) . There, for breadcrumb savings in CI runtime (say, run 1000 tests instead of 1005), a noticeable test organization/maintenance regression was made. Here, we say that running 1400 instead of 700 tests (100% more) is ok.

CI capacity is a non-issue. If we need more CI capacity we buy it.
If people want to keep arguing that letting stuff bit-rot to save time in CI is a good idea I don't know what to say. I don't know the history behind [11aa39b] and I had nothing to do with it, but if you have a valid claim about regression open a bug instead of complaining here.

**Review by jukkar — APPROVED** [review]
It seems that everybody more or less agrees with this PR and I am too.
I just want to point out that most of the issues that are seen if we remove the whitelisting, comes from the test application issues. So test app cannot be run because of too few net_buf's, or some config option is missing etc. I have not yet seen issue that indicates an issue in the core network stack itself.

**Comment by pfalcon** [comment]
@andrewboie 

> What does #12984 have to do with this? Did you mean to link something else? If the network stack has numerous issues on non-x86/ARM where are the open bugs?

Well, "network stack" is just one of many Zephyr subsystems, and there're various patterns how related issues affect various subsystems. I personally find it useful to think about such patterns, instead of applying local on-spot changes in a random fashion. In that regard, I find it interesting that subsystem X constrains testing too much, so issues go unspotted, while subsystem Y has too unconstrained testing, so it barely can be developed, as almost any change fails one of the less elaborated archs/targets (and anything else but x86/ARM is "less elaborated"). I'd personally be prepared to find compromise in the middle.

> We don't need to blacklist anything

Well, I personally still think that that SAM board with the misfitting driver should be blacklisted, and number of networking buffers should be set back to the (much lower) figures they had before it was added to CI run. So, mileage varies.

> I don't know the history behind [11aa39b] and I had nothing to do with it, but if you have a valid claim about regression open a bug instead of complaining here.

Oh, if only we had a common view/understanding of what's regression and what's not...


Bottom line: based on previous history, I personally can't be sure that this change is "right". (A minor point right here, as this one will be voted up, but there're many more "hard to decide" cases like this, and that's the concern.)

**Comment by andrewboie** [comment]
> Bottom line: based on previous history, I personally can't be sure that this change is "right". (A minor point right here, as this one will be voted up, but there're many more "hard to decide" cases like this, and that's the concern.)

Deciding what to test and how to go about it is certainly a conversation worth having.

However, adopting a policy that essentially reduces to "unlike other subsystems, we're just going to test network stuff on three board targets only, in a way that requires modifying dozens of testcase.yaml files to adjust" doesn't make much sense.

> Well, I personally still think that that SAM board with the misfitting driver should be blacklisted

If it's that bad, consider removing the "netif" tag from sam_e70_xplained's board description, it won't fall under the scope of net tests any more.

> I just want to point out that most of the issues that are seen if we remove the whitelisting, comes from the test application issues. So test app cannot be run because of too few net_buf's, or some config option is missing etc. I have not yet seen issue that indicates an issue in the core network stack itself.

What originally drove me to write this patch was when I introduced a special QEMU target to enable code coverage testing as part of routine CI builds #15886 and discovered that all the network tests were being skipped...and then found out I'd have to edit many testcase.yaml to get it going.

I get it, in most cases it's quite sufficient to prove that the arch-independent bits of the network stack (and also most every other subsystem in Zephyr) can have their functionality proven with a simple 'sanitycheck -p qemu_x86' but let's not treat networking as a special case. If we want to overhaul how kernel testing is done in general we need to take a more holistic view.

**Comment by pfalcon** [comment]
This fails CI with:

> ERROR: section .ramfunc is 128 bytes in the virtual/runtime address space, but only 0 in the loaded/XIP section!

And only for i.MXRT boards: mimxrt1064_evk, mimxrt1050_evk_qspi, mimxrt1050_evk.

Ping to @MaureenHelm and @agansari as maintainers, and random ping @ioannisg - IIRC, you were thorough reviewer of PR which introduced .ramfunc stuff, maybe you'd have an idea what's wrong here.

**Comment by ioannisg** [comment]
> 
> 
> This fails CI with:
> 
> > ERROR: section .ramfunc is 128 bytes in the virtual/runtime address space, but only 0 in the loaded/XIP section!
> 
> And only for i.MXRT boards: mimxrt1064_evk, mimxrt1050_evk_qspi, mimxrt1050_evk.
> 
> Ping to @MaureenHelm and @agansari as maintainers, and random ping @ioannisg - IIRC, you were thorough reviewer of PR which introduced .ramfunc stuff, maybe you'd have an idea what's wrong here.

Did some investigation on this one; I have a theory but I want @nashif  and @andyross to confirm it: when i build the test for `mimxrt1064_evk`, and using @andrewboie 's patch in `prj.conf`, the SRAM sections appear as follows:
```
0x80000000 - 0x800000080: NO CACHE Section 
0x80000080 - 0x80000080: RAMFUNC Section (zero length)
0x80000080 - 0x80000080: gap
0x80000100 - ... APP partitions
```

I believe that due to the gap, the check_link_map.py is failing, somehow, because it assumes that the gap is part of the RAMFUNC section, but nothing is actually loaded.

**Comment by agansari** [comment]
> Did some investigation on this one; I have a theory but I want @nashif and @andyross to confirm it: when i build the test for `mimxrt1064_evk`, and using @andrewboie 's patch in `prj.conf` (`CONFIG_NEWLIB_LIBC_ALIGNED_HEAP_SIZE=256`), the SRAM sections appear as follows:
> 
> ```
> 0x80000000 - 0x800000080: NO CACHE Section 
> 0x80000080 - 0x80000080: RAMFUNC Section (zero length)
> 0x80000080 - 0x80000080: gap
> 0x80000100 - ... APP partitions
> ```
> 
> I believe that due to the gap, the `check_link_map.py `is failing, somehow, because it assumes that the gap is part of the RAMFUNC section, but nothing is actually loaded.

@ioannisg I also suspect the script to be at fault here but for a different reason, i've built /zephyr/tests/net/socket/getaddrinfo/ and I get the message:
`ERROR: section .ramfunc is 128 bytes in the virtual/runtime address space, but only 0 in the loaded/XIP section!`
.nocache section is actually 128bytes (_nocache_ram_size = 0x80)
I suspect the script compares virtual .nocache (0x80) with loaded .ramfunc (0x0).

**Comment by ioannisg** [comment]
> 
> 
> > Did some investigation on this one; I have a theory but I want @nashif and @andyross to confirm it: when i build the test for `mimxrt1064_evk`, and using @andrewboie 's patch in `prj.conf` (`CONFIG_NEWLIB_LIBC_ALIGNED_HEAP_SIZE=256`), the SRAM sections appear as follows:
> > ```
> > 0x80000000 - 0x800000080: NO CACHE Section 
> > 0x80000080 - 0x80000080: RAMFUNC Section (zero length)
> > 0x80000080 - 0x80000080: gap
> > 0x80000100 - ... APP partitions
> > ```
> > 
> > 
> > I believe that due to the gap, the `check_link_map.py `is failing, somehow, because it assumes that the gap is part of the RAMFUNC section, but nothing is actually loaded.
> 
> @ioannisg I also suspect the script to be at fault here but for a different reason, i've built /zephyr/tests/net/socket/getaddrinfo/ and I get the message:
> `ERROR: section .ramfunc is 128 bytes in the virtual/runtime address space, but only 0 in the loaded/XIP section!`
> .nocache section is actually 128bytes (_nocache_ram_size = 0x80)
> I suspect the script compares virtual .nocache (0x80) with loaded .ramfunc (0x0).

@agansari  That's even funnier :) 
Yeah, it seems to me something is not OK in the script. I'll try to see if I can understand it, first, before attempting to fix it. @andyross , any available moment to take a look?

**Review by pfalcon — APPROVED** [review]

**Comment by marc-hb** [comment]
> Deciding what to test and how to go about it is certainly a conversation worth having.

Not using the word "whitelist" anywhere in that different and future conversation will go a long way. "Whitelist" means "don't ever use anything else", it never meant "default speed/coverage tradeoffs" and it never should.

> If we want to overhaul how kernel testing is done in general we need to take a more holistic view.

+1, is there already a github issue?

I think sanitycheck already does some kind of semi-random "platform sampling", still trying to find time to wrap my head around that part of the code. It also supports sharding.
